hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -254,64 +296,105 @@ private void handlePlatformTransaction(
-        // TODO: and have their own start/end. So system transactions are handled like separate user transactions.
+        // Commit all state changes
+        stack.commitFullStack();
 
         // store all records at once
         final var recordListResult = recordListBuilder.build();
+        recordCache.add(creator.nodeId(), payer, recordListResult.mainRecord().record(), consensusNow);
+        blockRecordManager.endUserTransaction(recordListResult.recordStream(), state);
+
+        return txBody;
+    }
+
+    @NonNull
+    private FeeAccumulator createFeeAccumulator(
+            @NonNull final SavepointStackImpl stack,
+            @NonNull final Configuration configuration,
+            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder) {
+        final var serviceApiFactory = new ServiceApiFactory(stack, configuration);
+        final var tokenApi = serviceApiFactory.getApi(TokenServiceApi.class);
+        return new FeeAccumulatorImpl(tokenApi, recordBuilder);
+    }
+
+    private PreCheckResult runPreChecks(
+            @NonNull final Instant consensusNow,
+            @NonNull final HandleContextVerifier verifier,
+            @NonNull final PreHandleResult preHandleResult) {
+        final var txBody = preHandleResult.txInfo().txBody();
 
-        if (preHandleResult != null) {
-            // FUTURE: This needs to be replaced by a proper implementation, as can be found in PR
-            // https://github.com/hashgraph/hedera-services/pull/7473
-            recordCache.add(
-                    0, preHandleResult.payer(), recordListResult.mainRecord().record(), consensusNow);
-        } else {
-            throw new IllegalStateException("pre handle result was null!");
+        // Check if pre-handle was successful
+        if (preHandleResult.status() != SO_FAR_SO_GOOD) {
+            return new PreCheckResult(preHandleResult.status(), preHandleResult.responseCode());
         }
 
-        blockRecordManager.endUserTransaction(recordListResult.recordStream(), state);
+        // Check the time box of the transaction
+        try {
+            checker.checkTimeBox(txBody, consensusNow);
+        } catch (final PreCheckException e) {
+            return new PreCheckResult(PRE_HANDLE_FAILURE, e.responseCode());
+        }
+
+        // Check all signature verifications. This will also wait, if validation is still ongoing.
+        final var payerKeyVerification = verifier.verificationFor(preHandleResult.payerKey());
+        if (payerKeyVerification.failed()) {
+            return new PreCheckResult(NODE_DUE_DILIGENCE_FAILURE, INVALID_SIGNATURE);
+        }
+
+        for (final var key : preHandleResult.requiredKeys()) {
+            final var verification = verifier.verificationFor(key);
+            if (verification.failed()) {
+                return new PreCheckResult(PRE_HANDLE_FAILURE, INVALID_SIGNATURE);
```

Comment:
same here why is this `PreCheckResult` ?

Reply:
I am not sure I understand the question. These are checks that we do before calling `handle()` of the Handler, i.e., pre-checks. Do you prefer a different name?

Reply:
I was wondering preCheckResult is dedicated to `PreHandleWorkflow` . May be I misunderstood

Reply:
Oh, I see. `PreCheckResult` is a record I created locally because I need to return two parameters: the error code and some indication if it is a due diligence failure. I reused PreHandleResult.Status because it was available, but we can do this differently if it makes the code easier to understand.

Reply:
yeah not urgent for this PR, it is somewhat confusing to return `PRE_HANDLE_FAILURE` and `PreCheckResult` in `HandleWorkflow`. We can clean up later also